cmd/gateway/s3x/gateway-s3x-ledger.go

```diff
@@ -15,418 +17,84 @@ Any claiming or releasing of locks should be done in the public setter+getter fu
-		Name:     objectName,
-		IpfsHash: objectHash,
-	}
-	return le.putLedger(ledger)
-}
-
-// DeleteBucket is used to remove a ledger bucket entry
-func (le *LedgerStore) DeleteBucket(name string) error {
-	le.Lock()
-	defer le.Unlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if ledger.GetBuckets()[name].Name == "" {
-		return ErrLedgerBucketDoesNotExist
-	}
-	delete(ledger.Buckets, name)
-	return le.putLedger(ledger)
-}
-
 // Close shuts down the ledger datastore
-func (le *LedgerStore) Close() error {
-	le.Lock()
-	defer le.Unlock()
-	return le.ds.Close()
+func (ls *ledgerStore) Close() error {
+	//todo: clean up caches
+	return ls.ds.Close()
 }
 
 /////////////////////
 // GETTER FUNCTINS //
 /////////////////////
 
-// GetObjectParts is used to return multipart upload parts
-func (le *LedgerStore) GetObjectParts(id string) ([]ObjectPartInfo, error) {
-	le.RLock()
-	defer le.RUnlock()
-	ledger, err := le.getLedger()
+// GetObjectInfos returns a list of ordered ObjectInfos with given prefix ordered by name
+func (ls *ledgerStore) GetObjectInfos(ctx context.Context, bucket, prefix, startsFrom string, max int) ([]ObjectInfo, error) {
+	defer ls.locker.read(bucket)()
+	b, err := ls.getBucketLoaded(ctx, bucket)
 	if err != nil {
 		return nil, err
 	}
-	if err := le.multipartExists(ledger, id); err != nil {
-		return nil, err
+	var names []string
+	objs := b.GetBucket().GetObjects()
+	for name := range objs {
+		if strings.HasPrefix(name, prefix) && strings.Compare(startsFrom, name) <= 0 {
+			names = append(names, name)
+		}
 	}
-	return ledger.GetMultipartUploads()[id].ObjectParts, nil
-}
-
-// MultipartIDExists is used to lookup if the given multipart id exists
-func (le *LedgerStore) MultipartIDExists(id string) error {
-	le.RLock()
-	defer le.RUnlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	return le.multipartExists(ledger, id)
-}
-
-// IsEmptyBucket checks if a bucket is empty or not
-func (le *LedgerStore) IsEmptyBucket(name string) error {
-	le.RLock()
-	defer le.RUnlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
-	}
-	if len(ledger.Buckets[name].Objects) == 0 {
-		return nil
-	}
-	return ErrLedgerNonEmptyBucket
-}
-
-// BucketExists is a public function to check if a bucket exists
-func (le *LedgerStore) BucketExists(name string) bool {
-	le.RLock()
-	defer le.RUnlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return false
+	sort.Strings(names)
+	if max > 0 && len(names) > max {
+		names = names[:max]
 	}
-	return le.bucketExists(ledger, name)
-}
-
-// ObjectExists is a public function to check if an object exists, and returns the reason
-// the object can't be found if any
-func (le *LedgerStore) ObjectExists(bucketName, objectName string) error {
-	le.RLock()
-	defer le.RUnlock()
-	ledger, err := le.getLedger()
-	if err != nil {
-		return err
+	list := make([]ObjectInfo, 0, len(names))
+	for _, name := range names {
```

Comment:
should do `for i, name := range names`, that way we can rewrite like this to avoid usage of append

```Golang
for i, name := range names {
  obj, err := ls.object(ctx, bucket, name)
  if err != nil {
     return nil, err
  }
  list[i] = obj.GetObjectInfo()
)
```

Reply:
Personally, I like the append better, and with the capacity set, both code should have the same performance.

Reply:
good point